### ISSUE TITLE
[minor] Linear8bitLt threading performance with multiple gpus

### ISSUE THREAD

**Issue by justheuristic** [TimDettmers/bitsandbytes#125]
__Premise:__ if you run LLM.8bit model with pytorch's threaded [DataParallel](https://pytorch.org/docs/stable/generated/torch.nn.DataParallel.html) or [TensorParallel](https://github.com/BlackSamorez/tensor_parallel), it would run much slower than on a single GPU. The quantized model would seemingly use only one GPU at a time.

Since i'm probably the only living being that uses threaded data parallelism, I'm gonna investigate this here and use this issue for updates. To: @TimDettmers , i've read contributing, but still not sure this is appropriate, if not - feel close the issue.

__Experiment setup:__ dual a100-80gb (sxm), torch 1.13.1+cu117, bitsandbytes 0.34 (reproduces in the latest version too), nvidia driver 510.39.01 . To isolate the issue, i'm running threads manually instead of through Data- or TensorParallel

```python
import time
import torch
from tqdm.auto import trange
import bitsandbytes as bnb
import bitsandbytes.functional as F
import threading
    
def benchmark(target, devices):
    print(f"Testing {target.__name__} on {devices}")
    t0 = time.perf_counter()
    threads = [threading.Thread(target=target, args=[device]) for device in devices]
    for thread in threads:
        thread.start()
    for thread in threads:
        thread.join()
    print(f"Benchmark for {target.__name__} on {devices} took {time.perf_counter() - t0:.3f} seconds\n")

    
def mlp_worker(device):
    with torch.cuda.device(device):
        b, d = 256, 4096
        m1 = bnb.nn.Linear8bitLt(d, 2 * d).to(torch.bfloat16).to(device)
        m2 = bnb.nn.Linear8bitLt(2 * d, d).to(torch.bfloat16).to(device)
        x = torch.randn(b, d, dtype=torch.bfloat16, device=device)
    
    with torch.inference_mode(), torch.cuda.device(device):
        for i in trange(1000):
            y = m2(torch.relu(m1(x))).sum()
    torch.cuda.synchronize(device)

benchmark(mlp_worker, ('cuda:0', 'cuda:1'))  # 7.7s
benchmark(mlp_worker, ('cuda:0',))  # 2.55s
benchmark(mlp_worker, ('cuda:1',))  # 2.55s

# note: i've re-run all benchmarks several times in the same session to account for cuda init time
```
![image](https://user-images.githubusercontent.com/3491902/213612641-32477eaf-9e09-43b3-a06d-a2c7121f81f5.png)

In other words, __running 2 gpus in parallel is so slower than it would take to run them sequentially__. The gpus seem to be interfering with one another in linear8bit forward. Notably, running the same code with regular `torch.nn.Linear` layers takes the same time in all cases (i.e. both GPUs work at full speed simultaneously), which is what you expect from DataParallel.
Removing one layer and relu does not change this behavior. Subjectively, the progress bars seem to be racing one another such that only one progressbar moves at a time.

__If you are reading this line and you have 5 minutes and two GPUs, please try and reproduce this on your machine and comment__

**Comment by justheuristic** [comment]
### What exactly is slowing it down?

Apparently, __igemmlt is not the culprit:__

```python
def igemmlt_worker(device):
    with torch.cuda.device(device):
        b, d = 256, 14336
        CxB = torch.randint(-127, 127, size=(d * 2, d), dtype=torch.int8, device=device)
        SB = (torch.Size([d * 2, d]), 'col_ampere')
        CA = torch.randint(-127, 127, size=(b, d), dtype=torch.int8, device=device)
        
    with torch.inference_mode(), torch.cuda.device(device):
        for i in trange(10_000):
            C32A, SA = F.transform(CA, "col32")
            out32, Sout32 = F.igemmlt(C32A, CxB, SA, SB)
    torch.cuda.synchronize(device)

benchmark(igemmlt_worker, ('cuda:0', 'cuda:1'))  # 5.4s
benchmark(igemmlt_worker, ('cuda:0',))  # 5.4s
benchmark(igemmlt_worker, ('cuda:1',))  # 5.4s
```

![image](https://user-images.githubusercontent.com/3491902/213613321-10f3d90b-4f02-47ab-93a2-830c1f7b7003.png)

**Comment by JubilantJerry** [comment]
Turns out you might not be the only one thinking about this issue. My suspicion is on this line of code:
https://github.com/TimDettmers/bitsandbytes/blob/25abf8d95f8a33f38e2ce6f637768b442379ccd9/bitsandbytes/functional.py#L459

This is setting the PyTorch CUDA context, and I think it causes global synchronization across all GPUs. Check if this `torch.cuda.set_device` alone accounts for your issue on your system as well, by adding it into your benchmark and removing the quantized Linear layers.
